### PR TITLE
Check for overflow before g_malloc() calls

### LINKED ISSUE #1533: memory out of bounds write in scard_process_list_readers
**version**

```
https://github.com/neutrinolabs/xrdp/blob/1c4e14415d923666ac60cb77c7e753ca24e0268d/sesman/chansrv/smartcard_pcsc.c#L600
```

**vuln**

`scard_process_list_readers` first read 4 bytes to `bytes_groups`, then use `g_malloc(bytes_groups + 1, 1)` to alloc memory.

if  `bytes_groups=0xffffffff` , it could trigger Integer overflow (`0xffffffff+1=0`), it could alloc 0 size memory to `groups`.

Then it could use `in_uint8a`  to copy data to  `groups`. resulting in **out-of-bounds write**

```
int
scard_process_list_readers(struct trans *con, struct stream *in_s)
{
    ..............................
    ..............................
    in_uint32_le(in_s, hContext);
    in_uint32_le(in_s, bytes_groups);
   
    groups = (char *) g_malloc(bytes_groups + 1, 1); // 0xffffffff + 1 = 0
    in_uint8a(in_s, groups, bytes_groups);  // copy 0xffffffff data to groups
```

### LINKED ISSUE #1534: memory out of bounds write in my_api_trans_data_in
**version**

```
https://github.com/neutrinolabs/xrdp/blob/1c4e14415d923666ac60cb77c7e753ca24e0268d/sesman/chansrv/chansrv.c#L1070
```

**vuln**

`my_api_trans_data_in` first read 4 bytes to `channel_name_bytes`, then use `g_new0(char, channel_name_bytes + 1)` to alloc memory.

if  `channel_name_bytes=0xffffffff` , it could trigger Integer overflow (`0xffffffff+1=0`), it could alloc 0 size memory to `chan_name`.

Then it could use `in_uint8a`  to copy data to  `chan_name`. resulting in **out-of-bounds write**

```
int
my_api_trans_data_in(struct trans *trans)
{
    ..........................
    ..........................
    else if (trans->extra_flags == 1)
    {
        rv = 1;
        in_uint32_le(s, channel_name_bytes);
        chan_name = g_new0(char, channel_name_bytes + 1);  // Integer overflow
        if (chan_name == NULL)
        {
            return 1;
        }
        in_uint8a(s, chan_name, channel_name_bytes); // out of bounds write
```

### LINKED ISSUE #1544: Integer overflow in xrdp_bitmap_create
**version**

```
https://github.com/neutrinolabs/xrdp/blob/1c4e14415d923666ac60cb77c7e753ca24e0268d/xrdp/xrdp_bitmap.c#L126
```

**vuln**

`process_server_create_os_surface`  first read some data from stream， then it could pass argument to `server_create_os_surface`.

```
static int
process_server_create_os_surface(struct mod *mod, struct stream *s)
{
    int rv;
    int rdpid;
    int width;
    int height;

    in_uint32_le(s, rdpid);
    in_uint16_le(s, width);
    in_uint16_le(s, height);
    rv = mod->server_create_os_surface(mod, rdpid, width, height);
    return rv;
}
```

then it could finally call to xrdp_bitmap_create

```
/*****************************************************************************/
struct xrdp_bitmap *
xrdp_bitmap_create(int width, int height, int bpp,
                   int type, struct xrdp_wm *wm)
{
   	...........................
   	
    switch (bpp)
    {
        case 8:
            Bpp = 1;
            break;
        case 15:
            Bpp = 2;
            break;
        case 16:
            Bpp = 2;
            break;
    }
    
    if (self->type == WND_TYPE_BITMAP || self->type == WND_TYPE_IMAGE)
    {
    	// width, height come from user input data.
        self->data = (char *)g_malloc(width * height * Bpp, 0);
    }
```

`width` and `height` is come from data, the **max value** of `width * height * Bpp` is `0x1fffc0002`.

```
>>> hex(0xffff*0xffff*2)
'0x1fffc0002'
```

the first argument  type of `g_malloc` is `int` , which **max value** is `0xffffffff`, So the size value will be truncated.

then It could lead **out of bounds write** to `self->data` later.

### PR DESCRIPTION
**Pull request by matt335672**
Here's a better attempt to replace #1551 

Fixes  #1533, fixes #1534, fixes #1544

#1533 - Checked the supported strings for the group parameter (see here). They're all relatively small, so code restructured to just allow a short string here, and check for it.
#1534 - Similarly the allowed string for a virtual channel name is quite small here - see here.
#1544 - Checked for integer overflow explicitly when requesting buffer.

I've not attempted to change the argument to g_malloc() to an unsigned integer here - I think that needs more thought and a separate PR

### CODE REVIEW

**Comment by matt335672** [comment]
Just realised that the change to `chansrv.c` can be further improved. The call to `g_malloc()`/`g_free()` for the channel name has now been replaced with a simple fixed-size stack allocation.

File: sesman/chansrv/chansrv.c [hunk 1]
```diff
@@ -39,6 +39,9 @@
 #include "xrdp_sockets.h"
 #include "audin.h"
 
+#define CHANNEL_NAME_BYTES 7
```
Comment (metalefty) [inline sesman/chansrv/chansrv.c:42]: Isn't this defined in MS specification?

Comment (matt335672) [reply sesman/chansrv/chansrv.c]: You're quite right. There's no name for it, but it's in section 2.2.1.3.4.1 of MS-RDPBCGR. I've added the definition to the include file and referenced that. Let me know if/when you want me to squash.

File: sesman/chansrv/chansrv.c [hunk 2]
```diff
@@ -1042,9 +1046,13 @@ my_api_trans_data_in(struct trans *trans)
     int rv;
     int bytes;
     int ver;
-    int channel_name_bytes;
     struct chansrv_drdynvc_procs procs;
-    char *chan_name;
+    /*
+     * Name is limited to CHANNEL_NAME_LEN for an SVC, or MAX_PATH
+     * bytes for a DVC
+     */
```
Comment (metalefty) [inline sesman/chansrv/chansrv.c:1053]: Good comment, thanks.

File: sesman/chansrv/chansrv.c (hunk 1 above)
Comment (metalefty) [reply sesman/chansrv/chansrv.c]: Thanks!

**Comment by metalefty** [comment]
@matt335672 
BTW, I and Jay are willing to invite you to xrdp members. We really appreciate your work not only by code but also on issue triages and answering users' questions. What about that?

After you joined xrdp members, you'll be able to close/reopen/edit issues, editing labels (issue triage), merge PRs (but usually don't) by yourself. Thank you very much for your work on xrdp!

**Comment by matt335672** [comment]
Well, I'm honoured, truly.

In short, 'yes'. At the moment I've got time to work on the project, and if you're prepared to allow me to help in a more hands-on way that would be great.

Thank you.

**Comment by metalefty** [comment]
Anyway, you should join the gitter chatrooms.
* https://gitter.im/neutrinolabs/xrdp
* https://gitter.im/neutrinolabs/xrdp-questions

**Comment by metalefty** [comment]
@matt335672 You've already added to xrdp team. You should now able to close or label issues. Feel free to do that.

**Pull request merged**